dsmr/typed_client.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package dsmr
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/ava-labs/hypersdk/proto/pb/dsmr"
+)
+
+var (
+	_ Marshaler[*dsmr.GetChunkRequest, *dsmr.GetChunkResponse, []byte]                   = (*getChunkMarshaler)(nil)
+	_ Marshaler[*dsmr.GetChunkSignatureRequest, *dsmr.GetChunkSignatureResponse, []byte] = (*getChunkSignatureMarshaler)(nil)
+	_ Marshaler[[]byte, []byte, *dsmr.ChunkCertificateGossip]                            = (*chunkCertificateGossipMarshaler)(nil)
+)
+
+type Marshaler[T any, U any, V any] interface {
+	MarshalRequest(T) ([]byte, error)
+	UnmarshalResponse([]byte) (U, error)
+	MarshalGossip(V) ([]byte, error)
+}
+
+// TODO merge upstream into avalanchego
+type TypedClient[T any, U any, V any] struct {
+	client    *p2p.Client
+	marshaler Marshaler[T, U, V]
+}
+
+func (t *TypedClient[T, U, _]) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	request T,
+	onResponse func(ctx context.Context, nodeID ids.NodeID, response U, err error),
+) error {
+	onByteResponse := func(ctx context.Context, nodeID ids.NodeID, responseBytes []byte, err error) {
+		response, parseErr := t.marshaler.UnmarshalResponse(responseBytes)
+		if parseErr != nil {
+			// TODO how do we handle this?
+			return
+		}
```

Comment:
This was my first thoughts but the semantics just are different since previously the error passed from the callback was an error from the peer (either a timeout or some error they defined), but now it's going to also include errors that we generated during handling the response. This might be fine to do since the timeout is technically also generated by ourselves, but I just left it as a TODO to think about more later.